File: packages/motif-vite/src/core/fluent/shell/layout.ts
import type { Declaration, ParsedClass } from '../../shared.js';
import { decl, fluent, fluentTypography, tokenDecl, withTransition } from '../../shared.js';

export function resolveFluentShellLayout(parsed: ParsedClass): Declaration[] | null {
  switch (`${parsed.utility}:${parsed.value ?? ''}`) {
    case 'drawer:':
      return withTransition(
        [
          decl('display', 'flex'),
          decl('flex-direction', 'column'),
          decl('gap', '1rem'),
          decl('min-width', '16rem'),
          decl('width', '100%'),
          decl('align-self', 'stretch'),
          decl('overflow', 'hidden'),
          tokenDecl('color', fluent('color', 'text')),
          tokenDecl('background-color', fluent('color', 'panel')),
          tokenDecl('background-image', fluent('effect', 'panel-tint')),
          decl('background-blend-mode', 'screen'),
          tokenDecl('border', fluent('border', 'panel')),
          tokenDecl('border-radius', fluent('radius', 'lg')),
          tokenDecl('padding', fluent('space', 'panel-pad')),
          tokenDecl('box-shadow', fluent('shadow', 'panel')),
          decl('backdrop-filter', 'blur(24px) saturate(1.18)'),
        ],
        fluent('effect', 'interactive-transition'),
        fluent('motion', 'duration'),
        fluent('motion', 'easing'),
      );
    case 'segmented:button':
      return withTransition(
        [
          ...fluentTypography('label-size', 'label-weight'),
          decl('display', 'inline-flex'),
          decl('align-items', 'center'),
          decl('justify-content', 'center'),
          decl('min-height', '2.5rem'),
          tokenDecl('color', fluent('color', 'text')),
          tokenDecl('background-color', fluent('color', 'surface-alt')),
          tokenDecl('border', fluent('border', 'action-subtle')),
          tokenDecl('border-radius', fluent('radius', 'md')),
          tokenDecl('padding', fluent('space', 'action-pad')),
        ],
        fluent('effect', 'interactive-transition'),
        fluent('motion', 'duration'),
        fluent('motion', 'easing'),
      );
    case 'search:field':
      return withTransition(
        [
          ...fluentTypography('body-size', 'body-weight'),
          decl('min-height', '2.75rem'),
          decl('padding-left', '2.5rem'),
          tokenDecl('color', fluent('color', 'text')),
          tokenDecl('caret-color', fluent('color', 'primary')),
          tokenDecl('background-color', fluent('color', 'field')),
          tokenDecl('background-image', fluent('effect', 'field-tint')),
          tokenDecl('border', fluent('border', 'field')),
          tokenDecl('border-radius', fluent('radius', 'sm')),
          tokenDecl('padding', fluent('space', 'field-pad')),
          tokenDecl('box-shadow', fluent('shadow', 'field')),
        ],
        fluent('effect', 'interactive-transition'),
        fluent('motion', 'duration'),
        fluent('motion', 'easing'),
      );
    case 'breadcrumb:':
      return withTransition(
        [
          decl('display', 'inline-flex'),
          decl('align-items', 'center'),
          decl('flex-wrap', 'wrap'),
          decl('gap', '0.5rem'),
          tokenDecl('color', fluent('color', 'muted')),
        ],
        fluent('effect', 'transition'),
        fluent('motion', 'duration'),
        fluent('motion', 'easing'),
      );
    case 'breadcrumb:item':
      return [
        ...fluentTypography('label-size', 'label-weight'),
        decl('display', 'inline-flex'),
        decl('align-items', 'center'),
        tokenDecl('color', fluent('color', 'muted')),
      ];
    case 'avatar:':
      return withTransition(
        [
          decl('display', 'inline-grid'),
          decl('place-items', 'center'),
          decl('inline-size', '2.5rem'),
          decl('block-size', '2.5rem'),
          tokenDecl('color', fluent('color', 'text')),
          tokenDecl('background-color', fluent('color', 'surface-alt')),
          decl('border-radius', '999px'),
          tokenDecl('border', fluent('border', 'action-subtle')),
        ],
        fluent('effect', 'interactive-transition'),
        fluent('motion', 'duration'),
        fluent('motion', 'easing'),
      );
    case 'persona:':
      return withTransition(
        [
          decl('display', 'flex'),
          decl('align-items', 'center'),
          decl('gap', '0.75rem'),
          tokenDecl('color', fluent('color', 'text')),
          tokenDecl('background-color', fluent('color', 'surface-alt')),
          tokenDecl('border', fluent('border', 'action-subtle')),
          tokenDecl('border-radius', fluent('radius', 'md')),
          tokenDecl('padding', fluent('space', 'surface-pad-sm')),
        ],
        fluent('effect', 'interactive-transition'),
        fluent('motion', 'duration'),
        fluent('motion', 'easing'),
      );
    case 'sheet:':
      return withTransition(
        [
          tokenDecl('background-color', fluent('color', 'panel')),
          tokenDecl('background-image', fluent('effect', 'panel-tint')),
          decl('background-blend-mode', 'screen'),
          tokenDecl('color', fluent('color', 'text')),
          tokenDecl('border', fluent('border', 'panel')),
          tokenDecl('border-radius', fluent('radius', 'lg')),
          tokenDecl('padding', fluent('space', 'panel-pad')),
          tokenDecl('box-shadow', fluent('shadow', 'panel')),
          decl('backdrop-filter', 'blur(24px) saturate(1.2)'),
          decl('max-width', '28rem'),
          decl('width', '100%'),
          decl('align-self', 'stretch'),
          decl('margin-inline', 'auto'),
          decl('overflow', 'hidden'),
        ],
        fluent('effect', 'interactive-transition'),
        fluent('motion', 'duration'),
        fluent('motion', 'easing'),
      );
    case 'sheet:side':
      return withTransition(
        [
          tokenDecl('background-color', fluent('color', 'panel')),
          tokenDecl('background-image', fluent('effect', 'panel-tint')),
          decl('background-blend-mode', 'screen'),
          tokenDecl('color', fluent('color', 'text')),
          tokenDecl('border', fluent('border', 'panel')),
          tokenDecl('border-radius', fluent('radius', 'lg')),
          tokenDecl('padding', fluent('space', 'panel-pad')),
          tokenDecl('box-shadow', fluent('shadow', 'panel')),
          decl('backdrop-filter', 'blur(24px) saturate(1.2)'),
          decl('max-width', '28rem'),
          decl('width', '100%'),
          decl('margin-inline', 'auto'),
          decl('min-height', '100%'),
          decl('margin-right', '0'),
          decl('border-top-right-radius', '0'),
          decl('border-bottom-right-radius', '0'),
        ],
        fluent('effect', 'interactive-transition'),
        fluent('motion', 'duration'),
        fluent('motion', 'easing'),
      );
    case 'sheet:bottom':
      return withTransition(
        [
          tokenDecl('background-color', fluent('color', 'panel')),
          tokenDecl('background-image', fluent('effect', 'panel-tint')),
          decl('background-blend-mode', 'screen'),
          tokenDecl('color', fluent('color', 'text')),
          tokenDecl('border', fluent('border', 'panel')),
          tokenDecl('border-radius', fluent('radius', 'lg')),
          tokenDecl('padding', fluent('space', 'panel-pad')),
          tokenDecl('box-shadow', fluent('shadow', 'panel')),
          decl('backdrop-filter', 'blur(24px) saturate(1.2)'),
          decl('max-width', '36rem'),
          decl('width', '100%'),
          decl('margin-inline', 'auto'),
          decl('margin-bottom', '0'),
          decl('border-bottom-left-radius', '0'),
          decl('border-bottom-right-radius', '0'),
        ],
        fluent('effect', 'interactive-transition'),
        fluent('motion', 'duration'),
        fluent('motion', 'easing'),
      );
    default:
      return null;
  }
}

File: core/src/rule/fluent/shell/layout.rs
use super::super::super::shared::{append_search_field_adornment, append_transition};
use super::super::super::{declaration, token_declaration, Declaration};
use super::super::typography_from_tokens;
use crate::parse::ParsedClass;
use crate::token::TokenRegistry;

pub(super) fn resolve(parsed: &ParsedClass, tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    match (parsed.utility.as_str(), parsed.value.as_deref()) {
        ("dialog", None) => dialog(tokens),
        ("drawer", None) => drawer(tokens),
        ("segmented", Some("button")) => segmented_button(tokens),
        ("search", Some("field")) => search_field(tokens),
        ("breadcrumb", None) => breadcrumb(tokens),
        ("breadcrumb", Some("item")) => breadcrumb_item(tokens),
        ("avatar", None) => avatar(tokens),
        ("persona", None) => persona(tokens),
        ("sheet", None) => sheet(tokens),
        ("sheet", Some("side")) => sheet_side(tokens),
        ("sheet", Some("bottom")) => sheet_bottom(tokens),
        _ => None,
    }
}

fn dialog(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = vec![
        token_declaration("background-color", tokens.fluent.color.get("panel")?),
        token_declaration("background-image", tokens.fluent.effect.get("panel-tint")?),
        declaration("background-blend-mode", "screen"),
        token_declaration("color", tokens.fluent.color.get("text")?),
        token_declaration("border", tokens.fluent.border.get("panel")?),
        token_declaration("border-radius", tokens.fluent.radius.get("lg")?),
        token_declaration("padding", tokens.fluent.space.get("panel-pad")?),
        token_declaration("box-shadow", tokens.fluent.shadow.get("panel")?),
        declaration("backdrop-filter", "blur(24px) saturate(1.2)"),
        declaration("max-width", "32rem"),
        declaration("width", "100%"),
    ];
    append_transition(
        &mut declarations,
        tokens.fluent.effect.get("interactive-transition")?,
        tokens.fluent.motion.get("duration")?,
        tokens.fluent.motion.get("easing")?,
    );
    Some(declarations)
}
fn drawer(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = vec![
        declaration("display", "flex"),
        declaration("flex-direction", "column"),
        declaration("gap", "1rem"),
        declaration("min-width", "16rem"),
        declaration("width", "100%"),
        declaration("align-self", "stretch"),
        declaration("overflow", "hidden"),
        token_declaration("color", tokens.fluent.color.get("text")?),
        token_declaration("background-color", tokens.fluent.color.get("panel")?),
        token_declaration("background-image", tokens.fluent.effect.get("panel-tint")?),
        declaration("background-blend-mode", "screen"),
        token_declaration("border", tokens.fluent.border.get("panel")?),
        token_declaration("border-radius", tokens.fluent.radius.get("lg")?),
        token_declaration("padding", tokens.fluent.space.get("panel-pad")?),
        token_declaration("box-shadow", tokens.fluent.shadow.get("panel")?),
        declaration("backdrop-filter", "blur(24px) saturate(1.18)"),
    ];
    append_transition(
        &mut declarations,
        tokens.fluent.effect.get("interactive-transition")?,
        tokens.fluent.motion.get("duration")?,
        tokens.fluent.motion.get("easing")?,
    );
    Some(declarations)
}
fn segmented_button(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = typography_from_tokens(tokens, "label-size", "label-weight", None)?;
    declarations.extend([
        declaration("display", "inline-flex"),
        declaration("align-items", "center"),
        declaration("justify-content", "center"),
        declaration("min-height", "2.5rem"),
        token_declaration("color", tokens.fluent.color.get("text")?),
        token_declaration("background-color", tokens.fluent.color.get("surface-alt")?),
        token_declaration("border", tokens.fluent.border.get("action-subtle")?),
        token_declaration("border-radius", tokens.fluent.radius.get("md")?),
        token_declaration("padding", tokens.fluent.space.get("action-pad")?),
    ]);
    append_transition(
        &mut declarations,
        tokens.fluent.effect.get("interactive-transition")?,
        tokens.fluent.motion.get("duration")?,
        tokens.fluent.motion.get("easing")?,
    );
    Some(declarations)
}
fn search_field(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = typography_from_tokens(tokens, "body-size", "body-weight", None)?;
    declarations.extend([
        declaration("min-height", "2.75rem"),
        token_declaration("color", tokens.fluent.color.get("text")?),
        token_declaration("caret-color", tokens.fluent.color.get("primary")?),
        token_declaration("background-color", tokens.fluent.color.get("field")?),
        token_declaration("background-image", tokens.fluent.effect.get("field-tint")?),
        token_declaration("border", tokens.fluent.border.get("field")?),
        token_declaration("border-radius", tokens.fluent.radius.get("sm")?),
        token_declaration("padding", tokens.fluent.space.get("field-pad")?),
        token_declaration("box-shadow", tokens.fluent.shadow.get("field")?),
        declaration("backdrop-filter", "blur(14px) saturate(1.05)"),
    ]);
    append_search_field_adornment(&mut declarations);
    append_transition(
        &mut declarations,
        tokens.fluent.effect.get("interactive-transition")?,
        tokens.fluent.motion.get("duration")?,
        tokens.fluent.motion.get("easing")?,
    );
    Some(declarations)
}
fn breadcrumb(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = vec![
        declaration("display", "inline-flex"),
        declaration("align-items", "center"),
        declaration("flex-wrap", "wrap"),
        declaration("gap", "0.5rem"),
        token_declaration("color", tokens.fluent.color.get("muted")?),
    ];
    append_transition(
        &mut declarations,
        tokens.fluent.effect.get("transition")?,
        tokens.fluent.motion.get("duration")?,
        tokens.fluent.motion.get("easing")?,
    );
    Some(declarations)
}
fn breadcrumb_item(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = typography_from_tokens(tokens, "label-size", "label-weight", None)?;
    declarations.extend([
        declaration("display", "inline-flex"),
        declaration("align-items", "center"),
        token_declaration("color", tokens.fluent.color.get("muted")?),
    ]);
    Some(declarations)
}
fn avatar(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = vec![
        declaration("display", "inline-grid"),
        declaration("place-items", "center"),
        declaration("inline-size", "2.5rem"),
        declaration("block-size", "2.5rem"),
        token_declaration("color", tokens.fluent.color.get("text")?),
        token_declaration("background-color", tokens.fluent.color.get("surface-alt")?),
        declaration("border-radius", "999px"),
        token_declaration("border", tokens.fluent.border.get("action-subtle")?),
    ];
    append_transition(
        &mut declarations,
        tokens.fluent.effect.get("interactive-transition")?,
        tokens.fluent.motion.get("duration")?,
        tokens.fluent.motion.get("easing")?,
    );
    Some(declarations)
}
fn persona(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = vec![
        declaration("display", "flex"),
        declaration("align-items", "center"),
        declaration("gap", "0.75rem"),
        token_declaration("color", tokens.fluent.color.get("text")?),
        token_declaration("background-color", tokens.fluent.color.get("surface-alt")?),
        token_declaration("border", tokens.fluent.border.get("action-subtle")?),
        token_declaration("border-radius", tokens.fluent.radius.get("md")?),
        token_declaration("padding", tokens.fluent.space.get("surface-pad-sm")?),
    ];
    append_transition(
        &mut declarations,
        tokens.fluent.effect.get("interactive-transition")?,
        tokens.fluent.motion.get("duration")?,
        tokens.fluent.motion.get("easing")?,
    );
    Some(declarations)
}
fn sheet(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = dialog(tokens)?;
    declarations.retain(|decl| decl.property != "max-width");
    declarations.push(declaration("max-width", "28rem"));
    declarations.push(declaration("width", "100%"));
    declarations.push(declaration("align-self", "stretch"));
    declarations.push(declaration("margin-inline", "auto"));
    declarations.push(declaration("overflow", "hidden"));
    Some(declarations)
}
fn sheet_side(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = sheet(tokens)?;
    declarations.push(declaration("min-height", "100%"));
    declarations.push(declaration("margin-right", "0"));
    declarations.push(declaration("border-top-right-radius", "0"));
    declarations.push(declaration("border-bottom-right-radius", "0"));
    Some(declarations)
}
fn sheet_bottom(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = sheet(tokens)?;
    declarations.push(declaration("max-width", "36rem"));
    declarations.push(declaration("margin-bottom", "0"));
    declarations.push(declaration("width", "100%"));
    declarations.push(declaration("border-bottom-left-radius", "0"));
    declarations.push(declaration("border-bottom-right-radius", "0"));
    Some(declarations)
}
